app-connector/src/server/connection.ts

```diff
@@ -0,0 +1,194 @@
+    LOGGER.debug("Calling token URL '%s'", url)
+    return request({
+        uri: url,
+        method: "GET",
+        json: true,
+        rejectUnauthorized: !insecure,
+        resolveWithFullResponse: true
+    }).then((response: any) => {
+        if (response.statusCode !== 200) {
+            throw new Error("Calling token URL failed with status '" + response.statusCode + "' and body '" + JSON.stringify(response.body, null, 2) + "'")
+        }
+        LOGGER.debug("Token URL returned %s", JSON.stringify(response.body, null, 2))
+        return response.body
+    })
+}
+
+async function callCSRUrl(csrUrl: any, csr: any, insecure: any) {
+    LOGGER.debug("Calling csr URL '%s'", csrUrl)
+    var csrData = forge.util.encode64(csr)
+
+    return request.post({
+        uri: csrUrl,
+        body: { csr: csrData },
+        json: true,
+        rejectUnauthorized: !insecure,
+        resolveWithFullResponse: true
+    }).then((response: any) => {
+        if (response.statusCode !== 201) {
+            throw new Error("Calling CSR URL failed with status '" + response.statusCode + "' and body '" + JSON.stringify(response.body, null, 2) + "'")
+        }
+        LOGGER.debug("CSR URL returned")
+        return Buffer.from(response.body.crt, 'base64').toString("ascii")
+    })
+}
+
+async function callInfoUrl(infoUrl: any, crt: any, privateKey: any, insecure: any) {
+    LOGGER.debug("Calling info URL '%s'", infoUrl)
+
+    return request.get({
+        uri: infoUrl,
+        json: true,
+        agentOptions: {
+            cert: crt,
+            key: privateKey
+        },
+        rejectUnauthorized: !insecure,
+        resolveWithFullResponse: true
+    }).then((response: any) => {
+        if (response.statusCode !== 200) {
+            throw new Error("Calling Info URL failed with status '" + response.statusCode + "' and body '" + JSON.stringify(response.body, null, 2) + "'")
+        }
+        LOGGER.debug("Got following Info URL returned: %s", JSON.stringify(response.body, null, 2))
+        return response.body
+    })
+}
+function generateCSR(subject: any) {
+    LOGGER.debug("Creating CSR using subject %s", subject)
+    var pk = forge.pki.privateKeyFromPem(privateKey())
+    var publickey = forge.pki.setRsaPublicKey(pk.n, pk.e)
+
+    // create a certification request (CSR)
+    var csr = forge.pki.createCertificationRequest()
+    csr.publicKey = publickey
+
+    csr.setSubject(parseSubjectToJsonArray(subject))
+    csr.sign(pk)
+    LOGGER.debug("Created csr using subject %s", subject)
+    return forge.pki.certificationRequestToPem(csr)
+}
+
+function parseSubjectToJsonArray(subject: any) {
+    var subjectsArray: any = []
+    subject.split(",").map((el: any) => {
+        const val = el.split("=")
+        subjectsArray.push({
+            shortName: val[0],
+            value: val[1]
+        })
+    })
+
+    return subjectsArray;
+}
+
+function certificate() {
+    if (fs.existsSync(crtFile)) {

```

Comment:
That should not be file based. The code before was based only on in-memory data and read/wrote it once to file.

Reply:
made it an option